File: codechef/NOV13/PRETNUM/pretnum_1.py
N = 999999
prime = [0]*N
prime[0] = 1
prime[1] = 1
pret = [1]*N

def sieve():
    i = 2
    while i*i <= N:
        j = 2*i
        if prime[i] != 1:
            while j < N:
                prime[j] = 1
                j += i
        i += 1
##    for i in xrange(N):
##        if prime[i] == 0:
##            prime[i] = i
    for t in xrange(input()):
        m,n = map(long,raw_input().split())
        ourRange = []
        for i in xrange(m,n+1):
            ourRange.append(i)
        i = 2
        while i*i <= n:
            fr = m/i
            fr *= i
            j = fr
            while j <= n:
                if j!=i and j >= m:
                    ourRange[j-m] = 0
                j += i
            i += 1
        
        i = 2
        final = 0
        while i < N:
            if prime[i] != 1:
                j = i
                k = 1
                while j < m:
                    j = j*i
                    k += 1
                while j <= n:
                    if prime[k+1] == 0:
                        ourRange[j-m] = j
                        #final += 1
                    else:
                        ourRange[j-m] = 0
                    j = j*i
                    k += 1
            i += 1
        if ourRange[0] == 1:
            ourRange[0] = 0
        for i in xrange(m,n+1):
            if ourRange[i-m]:
                final += 1
        #print prime
        #print pret
        print final
sieve()

File: codechef/NOV13/PRETNUM/pretnum_1.cpp
#include<iostream>
#include<cstdio>
#define UPTO 999999
using namespace std;
int prime[1000000] = {0};
void sieve()
{
    prime[0] = 1;
    prime[1] = 1;
    int i = 2;
    int j;
    while (i*i <= UPTO)
    {
        j = 2*i;
        if (prime[i] != 1)
        {
            while(j<UPTO)
            {
                prime[j] = 1;
                j += i;
            }
        }
        i += 1;
    }
}
int main()
{
    sieve();
    int t;
    scanf(" %d",&t);
    while(t--)
    {
        long long i,j,m,n,from,k;
        long long ourRange[1000001];

        scanf(" %lld %lld",&m,&n);
        long long z;
        for(i=m,z=0;i<=n;i++,z++)
        {
            ourRange[z] = i;
        }
        for(i=2;i*i<n;i++)
        {
            from = m/i;
            from*=i;
            for(j=from;j<=n;j=j+i)
            {
                if(j!=i && j>=m)
                ourRange[j-m] = 0;
            }
        }
        if(m==1)
        {
            ourRange[0] = 0;
        }
        i = 2;
        long long res = 0;
        while(i<UPTO)
        {
            if (prime[i] == 0)
            {
                j = i;
                k = 1;
                while (j < m)
                {
                    j = j*i;
                    k += 1;
                }
                while(j<=n)
                {
                    if(prime[k+1] == 0)
                    {
                        ourRange[j-m] = j;
                    }
                    else
                    {
                        ourRange[j-m] = 0;
                    }
                    j = j*i;
                    k += 1;
                }
            }
            i += 1;
        }
        for(i=m;i<=n;i++)
        {
            if(ourRange[i-m])
            {
                res++;
            }
        }
        printf("%lld\n",res);
    }
    return 0;
}
